port/osx/omrosbacktrace_impl.c

```diff
@@ -0,0 +1,413 @@
+			portLibrary->mem_free_memory(portLibrary, tmp);
+		}
+	}
+
+	return i - discard;
+}
+
+/**
+ * This function takes a thread structure already populated with a backtrace by omrintrospect_backtrace_thread
+ * and looks up the symbols for the frames. The format of the string generated is:
+ * 		symbol_name (statement_id instruction_pointer [module+offset])
+ * If it isn't possible to determine any of the items in the string then they are omitted. If no heap is specified
+ * then this function will use malloc to allocate the memory necessary for the symbols which must be freed by the caller.
+ *
+ * @param portLibrary a pointer to an initialized port library
+ * @param threadInfo a thread structure populated with a backtrace
+ * @param heap a heap from which to allocate any necessary memory. If NULL malloc is used instead.
+ *
+ * @return the number of frames for which a symbol was constructed.
+ */
+uintptr_t
+omrintrospect_backtrace_symbols_raw(OMRPortLibrary *portLibrary, J9PlatformThread *threadInfo, J9Heap *heap)
+{
+	J9PlatformStackFrame *frame = NULL;
+	uintptr_t i = 0;
+
+	for (i = 0, frame = threadInfo->callstack; NULL != frame; frame = frame->parent_frame, i++) {
+		char output_buf[512];
+		char *cursor = output_buf;
+		Dl_info dlInfo;
+		uintptr_t length = 0;
+		uintptr_t iar = frame->instruction_pointer;
+		uintptr_t symbol_offset = 0;
+		uintptr_t module_offset = 0;
+		const char *symbol_name = "";
+		const char *module_name = "<unknown>";
+		size_t symbol_length = 0;
+
+		memset(&dlInfo, 0, sizeof(Dl_info));
+
+		/* do the symbol resolution while we're here */
+		if (0 != dladdr((void *)iar, &dlInfo)) {
+			if (NULL != dlInfo.dli_sname) {
+				uintptr_t sym = (uintptr_t)dlInfo.dli_saddr;
+				symbol_name = dlInfo.dli_sname;
+				symbol_length = strlen(symbol_name);
+				symbol_offset = iar - sym;
+			}
+
+			if (NULL != dlInfo.dli_fname) {
+				/* strip off the full path if possible */
+				module_name = strrchr(dlInfo.dli_fname, '/');
+				if (NULL != module_name) {
+					module_name += 1;
+				} else {
+					module_name = dlInfo.dli_fname;
+				}
+			}
+
+			if (NULL != dlInfo.dli_fbase) {
+				/* set the module offset */
+				module_offset = iar - (uintptr_t)dlInfo.dli_fbase;
+			}
+		}
+
+		/* symbol_name+offset (id, instruction_pointer [module+offset]) */
+		if (symbol_length > 0) {
+			cursor += omrstr_printf(portLibrary, cursor, sizeof(output_buf) - (cursor - output_buf), "%.*s", symbol_length, symbol_name);
+			cursor += omrstr_printf(portLibrary, cursor, sizeof(output_buf) - (cursor - output_buf), "+0x%x ", symbol_offset);
+		}
+		cursor += omrstr_printf(portLibrary, cursor, sizeof(output_buf) - (cursor - output_buf), "(0x%p", frame->instruction_pointer);
+		if ('\0' != module_name[0]) {
+			cursor += omrstr_printf(portLibrary, cursor, sizeof(output_buf) - (cursor - output_buf), " [%s+0x%x]", module_name, module_offset);
+		}
+		cursor += omrstr_printf(portLibrary, cursor, sizeof(output_buf) - (cursor - output_buf), ")\0");
```

Comment:
Why the explicit NUL (`\0`)? I think it can safely be removed.

Reply:
Doesn't the string need a NUL when it is written? The length will be calculated with `strlen`

Reply:
Actually my bad, forgot about the implicit NUL terminator